content/geometry/SegmentIntersection.h

```diff
@@ -1,53 +1,46 @@
 /**
- * Author: Ulf Lundstrom
- * Date: 2009-03-21
+ * Author: Victor Lecomte, chilli
+ * Date: 2019-04-27
  * License: CC0
- * Source:
+ * Source: https://vlecomte.github.io/cp-geo.pdf
  * Description:\\
 \begin{minipage}{75mm}
-If a unique intersetion point between the line segments going from s1 to e1 and from s2 to e2 exists r1 is set to this point and 1 is returned.
-If no intersection point exists 0 is returned and if infinitely many exists 2 is returned and r1 and r2 are set to the two ends of the common line.
+If a unique intersection point between the line segments going from s1 to e1 and from s2 to e2 exists then it is returned.
+If no intersection point exists an empty vector is returned. If infinitely many exist a vector with 2 elements is returned.
 The wrong position will be returned if P is Point<int> and the intersection point does not have integer coordinates.
 Products of three coordinates are used in intermediate steps so watch out for overflow if using int or long long.
-Use segmentIntersectionQ to get just a true/false answer.
 \end{minipage}
 \begin{minipage}{15mm}
 \includegraphics[width=\textwidth]{content/geometry/SegmentIntersection}
 \end{minipage}
- * Status: Well tested with unitTest and with Kattis problem intersection.
+ * Status: Well tested with fuzz-test and with Kattis problem intersection.
  * Usage:
- * Point<double> intersection, dummy;
- * if (segmentIntersection(s1,e1,s2,e2,intersection,dummy)==1)
- *   cout << "segments intersect at " << intersection << endl;
+ * vector<P> inter = segInter(s1,e1,s2,e2);
+ * if (inter.size()==1)
+ *   cout << "segments intersect at " << inter[0] << endl;
  */
 #pragma once
 
 #include "Point.h"
+#include "onSegment.h"
 
-template<class P>
-int segmentIntersection(const P& s1, const P& e1,
-		const P& s2, const P& e2, P& r1, P& r2) {
-	if (e1==s1) {
-		if (e2==s2) {
-			if (e1==e2) { r1 = e1; return 1; } //all equal
-			else return 0; //different point segments
-		} else return segmentIntersection(s2,e2,s1,e1,r1,r2);//swap
+template <class P>
+bool segInterProper(P a, P b, P c, P d, P& out) {
+	double oa = c.cross(d, a), ob = c.cross(d, b),
+	       oc = a.cross(b, c), od = a.cross(b, d);
+	if (sgn(oa) * sgn(ob) < 0 && sgn(oc) * sgn(od) < 0) {
+		out = (a * ob - b * oa) / (ob - oa);
```

Comment:
(very nice to see the returned value become just a division without any further translations -- then the result is precise when inputs are integers, so you can put intersections in sets etc.)